src/scheduler.ml

```diff
@@ -140,125 +218,147 @@ end = struct
 
     let rec buffer_thread () =
       Mutex.lock event_mtx;
-      if List.is_empty !new_events then
+      if List.is_empty !files_changed then
         Condition.wait event_cv event_mtx;
-      let events_to_send = !new_events in
-      new_events := [];
+      let files = !files_changed in
+      files_changed := [];
       Mutex.unlock event_mtx;
-      List.iter events_to_send ~f:(fun batch ->
-        Event.(sync (send chan batch)));
+      Event.send_files_changed files;
       Thread.delay buffering_time;
       buffer_thread ()
     in
 
-    ignore (Thread.create worker_thread ());
-    buffer_thread ()
+    ignore (Thread.create worker_thread () : Thread.t);
+    ignore (Thread.create buffer_thread () : Thread.t))
 
-  let should_trigger_rebuild fns =
-    List.filter fns ~f:(fun fn ->
-      let fn = Path.to_absolute_filename fn in
-      if String.Table.mem ignored_files fn then begin
-        (* only use ignored record once *)
-        String.Table.remove ignored_files fn;
-        false
-      end
-      else true)
-    |> List.is_empty
-    |> not
+  let init () = Lazy.force init
 end
 
-let ignore_for_watch = Watch.ignore_file
+module Process_watcher : sig
+  (** Initialize the process watcher thread. *)
+  val init : unit -> unit
 
-module Running_jobs : sig
-  val add : running_job -> unit
-  val count : unit -> int
+  (** Register a new running job. *)
+  val register_job : job -> unit
 
-  (** Runs forever, communicating events to channel when children terminate. *)
-  val wait_to_chan : (int * Unix.process_status) Event.channel -> 'a
-
-  (** Removes a process previously returned via channel from the table. *)
-  val resolve_and_remove_job : int -> running_job
-
-  (** Send the following signal to all running jobs *)
+  (** Send the following signal to all running processes. *)
   val killall : int -> unit
 end = struct
-  let all = Hashtbl.create 128
-
-  let running_pids = ref Int.Set.empty
-  let running_pids_mtx = Mutex.create ()
+  let mutex = Mutex.create ()
   let something_is_running_cv = Condition.create ()
 
-  let add job =
-    Hashtbl.add all job.pid job;
-    Mutex.lock running_pids_mtx;
-    running_pids := Int.Set.add !running_pids job.pid;
-    if Int.Set.cardinal !running_pids = 1 then
-      Condition.signal something_is_running_cv;
-    Mutex.unlock running_pids_mtx
-
-  let resolve_and_remove_job pid =
-    let job =
-      match Hashtbl.find all pid with
-      | Some job -> job
-      | None -> assert false
-    in
-    Hashtbl.remove all pid;
-    job
+  module Process_table : sig
+    val add : job -> unit
+    val remove : pid:int -> Unix.process_status -> unit
+    val running_count : unit -> int
+    val iter : f:(job -> unit) -> unit
+  end = struct
+    type process_state =
+      | Running of job
+      | Zombie of Unix.process_status
+
+    let table = Hashtbl.create 128
+    let running_count = ref 0
+
+    let _invariant () =
```

Comment:
I think that this should be checked (in an assert) or removed.

Reply:
I made it a comment